src/hotspot/share/oops/compressedOops.cpp

```diff
@@ -186,83 +186,60 @@ NarrowPtrStruct CompressedKlassPointers::_narrow_klass = { nullptr, 0, true };
 //  are compiled for 32bit to LP64_ONLY).
 size_t CompressedKlassPointers::_range = 0;
 
+#ifdef _LP64
+
+// Given a klass range [addr, addr+len) and a given encoding scheme, assert that this scheme covers the range, then
+// set this encoding scheme. Used by CDS at runtime to re-instate the scheme used to pre-compute klass ids for
+// archived heap objects.
+void CompressedKlassPointers::initialize_for_given_encoding(address addr, size_t len, address requested_base, int requested_shift) {
+  assert(is_valid_base(requested_base), "Address must be a valid encoding base");
+  address const end = addr + len;
+
+  const int narrow_klasspointer_bits = sizeof(narrowKlass) * 8;
+  const size_t encoding_range_size = nth_bit(narrow_klasspointer_bits + requested_shift);
+  address encoding_range_end = requested_base + encoding_range_size;
+
+  assert(requested_base <= addr && encoding_range_end >= end, "Encoding does not cover the full Klass range");
```

Comment:
Is it possible for `requested_base` to ever be less than `addr`?

Reply:
Not at the moment, no. Probably never. I separated those two mainly for cleanliness, since Klass range and encoding range are different things.

Reply:
I am fine having them separately but that assert condition forced me to think if requested_base can ever be less than addr.  If not, then I guess assert for `requested_base == addr` ?

Reply:
Sure, makes sense.